### DISCUSSION TITLE
Register logger in container (log4net)

### DISCUSSION THREAD

**Discussion by ErwinDraconis** [PrismLibrary/Prism#2335] [WPF] [answered]
Hello,
 i want to register a logger in the container, i use log4net, its working but the problem is that the log context is always 'App'

![image](https://user-images.githubusercontent.com/42243889/107069265-ffb0a000-67e1-11eb-983a-9673a7cbc02e.png)

injection in the view model

![image](https://user-images.githubusercontent.com/42243889/107069294-0c34f880-67e2-11eb-84fe-f2f0da9569cb.png)

the output in text file is always pointing to App instead of the curent context which is in this case the viewModel

![image](https://user-images.githubusercontent.com/42243889/107069331-19ea7e00-67e2-11eb-94a8-5bcba4e3d5a0.png)

can you please tell me how to register the logger and change the context each time a class call for it ?

Thank you.

**Comment by ErwinDraconis** [comment]
Hello guys,
is there a way to inject the logger, instead of doing something like :
![image](https://user-images.githubusercontent.com/42243889/107154686-aaf15e80-6974-11eb-98e3-2cecf51c0aed.png)

and in each class we instanciate a new instance of the logger.
Thank you

**Reply by Pretasoc** [reply]
I don't know any, which will just operate on the Prism interfaces. I know of a [Solution](https://github.com/unitycontainer/log4net) for unity container, that will resolve the correct type. Other containers will probably support something like that as well.

**Reply by ErwinDraconis** [reply]
@Pretasoc Thank you for your feedback, i installed the package Unity.log4net and register the logger as follow : 
please note its the way that PRISM 7 register types , i dont know if they use unity underhoods, and there is no AddNewExtension method in the containerRegistry class, so i process with Register  

`containerRegistry.Register<Log4NetExtension>();`

but i still get the same error : 
 
![image](https://user-images.githubusercontent.com/42243889/107229807-8c47a200-6a1e-11eb-90e9-9b47190c953c.png)

**Comment by dansiegel** [comment, marked as answer]
Other than the use of Open Generics, Prism's abstractions don't really have a way of handling the more complex factory that would be required to produce an instance of ILog with the name of the type that it is injected into. Not to say that it can't be done, but you will have to look to the underlying container for how to more appropriately handle this situation.

As was already suggested you might try Unity.Log4Net if you are using Unity... If you're using DryIoc you may look to this sample from @dadhi on how to configure the container with Serilog. 

https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/ExamplesContextBasedResolution.md#serilog-logger

**Reply by ErwinDraconis** [reply]
@dansiegel thank you so much for the help, the solution was a mix of all the answers, i will post it below

**Comment by brianlagunas** [comment]
I'll also add that you can always access the underlying container directly by using the `containerRegistry.GetContainer()` method.

**Reply by ErwinDraconis** [reply]
@brianlagunas thanks a lot, this was the missing piece.

**Comment by ErwinDraconis** [comment]
Hello, 
hier is the solution i come up with, based on all suggestion above : 
first install Unity.log4net package, then register it like below : 

`containerRegistry.GetContainer().AddNewExtension<Log4NetExtension>();`

now you are able to inject ILog in your view model : 

![image](https://user-images.githubusercontent.com/42243889/107330875-708ddb80-6ab2-11eb-90a5-517218f648f1.png)

Thank you.
